src/qibo/optimizers/gradient_based.py

```diff
@@ -1 +1,135 @@
+                <https://www.tensorflow.org/api_docs/python/tf/keras/optimizers>_
+                for the list of available optimizers.
+            options (dict): options which can be provided to the chosen optimizer.
+                See the same reference of above for the complete list of options once
+                the optimizer is selected.
+
+        Example:
+        .. testcode::
+            import numpy as np
+            from qibo import models, hamiltonians, gates, set_backend
+            from qibo.optimizers.gradient_based import TensorFlowSGD
+
+            # tensorflow backend is needed to use the TensorFlowSGD optimizer.
+            set_backend("tensorflow")
+
+            # define a dummy model
+            nqubits = 2
+            nlayers = 3
+
+            c = models.Circuit(nqubits)
+            for l in range(nlayers):
+                for q in range(nqubits):
+                    c.add(gates.RY(q=q, theta=0))
+                    c.add(gates.RY(q=q, theta=0))
+                for q in range(nqubits-1):
+                    c.add(gates.CNOT(q0=q, q1=q+1))
+            c.add(gates.M(*range(nqubits)))
+
+            # define a loss function
+            h = hamiltonians.Z(nqubits)
+            def loss(parameters, circuit, hamiltonian):
+                circuit.set_parameters(parameters)
+                return hamiltonian.expectation(circuit().state())
+
+            # initialize parameters
+            params = np.random.randn(2 * nqubits * nlayers)
+
+            # initialize optimizer
+            options = {"learning_rate": 0.05}
+            opt = TensorFlowSGD(initial_parameters=params, loss=loss, args=(c,h), options=options)
+            # perform the training
+            res = opt.fit(epochs=50, nmessage=1)
+        """
+
+        super().__init__(initial_parameters, args, loss)
+        # This optimizer works only with tensorflow backend
+        self.backend = TensorflowBackend()
+
+        self.options = options
+
+        # options are automatically checked inside the tf.keras.optimizer
+        self.optimizer = getattr(self.backend.tf.optimizers, optimizer)(**self.options)
+
+    def fit(self, epochs=100000, nmessage=100, loss_treshold=None):
+        """
+        Compute the SGD optimization according to the chosen optimizer.
+
+        Args:
+            epochs (int): number of optimization iterations [default 10000].
+            nmessage (int): Every how many epochs to print
+                a message of the loss function [default 100].
+            loss_treshold (float): if this loss function value is reached, training
+                stops [default None].
+
+        Returns:
+            (float): best loss value
+            (np.ndarray): best parameter values
+            (list): loss function history
+        """
+
+        vparams = self.backend.tf.Variable(
+            self.params, dtype=self.backend.tf.complex128
+        )
+        loss_history = []
+
+        def sgd_step():
```

Comment:
Is there a particular reason why we don't want to expose the sgd step to the user? I think in some cases it could be useful to be able to define the training loop your self for customization.

Reply:
I was only keeping the notation we had in the old optimizer file. 
I am open to move it outside.